File: ts/src/escape.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */


// The variant names, in the order the reference lists them. `none` is
// not here: it is the absent argument, and the absent argument is the
// C/JSON convention.
const ESC_VARIANTS = ['sq', 'sql', 'shell', 'xml', 'uri', 'regex']


const ESC_REGEX_PUNCT = '\\.+*?()[]{}|^$/'


const ESC_XML: [string, string][] = [
  ['&', '&amp;'],
  ['<', '&lt;'],
  ['>', '&gt;'],
  ['"', '&quot;'],
  ["'", '&apos;'],
]


const ESC_URI_UNRESERVED =
  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~'


function isEscVariant(v: string): boolean {
  return ESC_VARIANTS.includes(v)
}


// The C escape, in whichever quote the convention uses. Control
// characters below U+0020 that have no short form become `\uXXXX` with
// LOWERCASE hex, as JSON writes them; U+007F is not a JSON escape and
// is left alone.
function escC(src: string, quote: string): string {
  let out = ''
  for (const ch of src) {
    if ('\\' === ch) { out += '\\\\'; continue }
    if (quote === ch) { out += '\\' + quote; continue }
    if ('\n' === ch) { out += '\\n'; continue }
    if ('\r' === ch) { out += '\\r'; continue }
    if ('\t' === ch) { out += '\\t'; continue }
    if ('\b' === ch) { out += '\\b'; continue }
    if ('\f' === ch) { out += '\\f'; continue }
    const c = ch.codePointAt(0) as number
    if (0x20 > c) {
      out += '\\u' + c.toString(16).padStart(4, '0')
      continue
    }
    out += ch
  }
  return out
}


// The inverse. A `\uXXXX` naming half a surrogate pair is read WITH its
// partner or refused: a lone surrogate is not a character either port
// can carry, and answering one in TypeScript where Go cannot is the
// divergence the whole table exists to avoid.
function unescC(src: string, quote: string): [string, boolean] {
  const bad: [string, boolean] = ['', false]
  let out = ''

  for (let i = 0; i < src.length; i++) {
    const ch = src[i]
    if ('\\' !== ch) { out += ch; continue }

    i++
    if (src.length <= i) { return bad }
    const n = src[i]

    if ('\\' === n) { out += '\\'; continue }
    if (quote === n) { out += quote; continue }
    if ('n' === n) { out += '\n'; continue }
    if ('r' === n) { out += '\r'; continue }
    if ('t' === n) { out += '\t'; continue }
    if ('b' === n) { out += '\b'; continue }
    if ('f' === n) { out += '\f'; continue }

    if ('u' !== n) { return bad }

    const hi = hex4(src, i + 1)
    if (0 > hi) { return bad }
    i += 4

    if (0xDC00 <= hi && 0xDFFF >= hi) { return bad }

    if (0xD800 <= hi && 0xDBFF >= hi) {
      // A high surrogate must be followed by its own `\u` low one.
      if ('\\' !== src[i + 1] || 'u' !== src[i + 2]) { return bad }
      const lo = hex4(src, i + 3)
      if (0 > lo || 0xDC00 > lo || 0xDFFF < lo) { return bad }
      i += 6
      out += String.fromCharCode(hi, lo)
      continue
    }

    out += String.fromCharCode(hi)
  }

  return [out, true]
}


function hex4(src: string, at: number): number {
  if (src.length < at + 4) { return -1 }
  let v = 0
  for (let i = at; i < at + 4; i++) {
    const d = hexDigit(src[i])
    if (0 > d) { return -1 }
    v = v * 16 + d
  }
  return v
}


function hexDigit(ch: string): number {
  if ('0' <= ch && '9' >= ch) { return ch.charCodeAt(0) - 0x30 }
  if ('a' <= ch && 'f' >= ch) { return ch.charCodeAt(0) - 0x61 + 10 }
  if ('A' <= ch && 'F' >= ch) { return ch.charCodeAt(0) - 0x41 + 10 }
  return -1
}


// Standard SQL: a literal's only escape is the doubled quote. A LONE
// quote has no inverse -- it is the escape character, so its meaning
// depends on what was meant to follow it.
function escSql(src: string): string {
  return src.split("'").join("''")
}

function unescSql(src: string): [string, boolean] {
  let out = ''
  for (let i = 0; i < src.length; i++) {
    if ("'" !== src[i]) { out += src[i]; continue }
    if ("'" !== src[i + 1]) { return ['', false] }
    out += "'"
    i++
  }
  return [out, true]
}


// POSIX single-quoting: the caller wraps the value in single quotes,
// and a quote inside it closes, escapes and reopens them.
function escShell(src: string): string {
  return src.split("'").join("'\\''")
}

function unescShell(src: string): [string, boolean] {
  let out = ''
  for (let i = 0; i < src.length; i++) {
    if ("'" !== src[i]) { out += src[i]; continue }
    if ("\\" !== src[i + 1] || "'" !== src[i + 2] || "'" !== src[i + 3]) {
      return ['', false]
    }
    out += "'"
    i += 3
  }
  return [out, true]
}


function escXml(src: string): string {
  let out = ''
  for (const ch of src) {
    const ent = ESC_XML.find((e) => e[0] === ch)
    out += undefined === ent ? ch : ent[1]
  }
  return out
}


// A bare `&` is REFUSED rather than passed through: it is the one
// character whose meaning is ambiguous here, since it opens an entity.
// Every other character is unambiguous and passes.
function unescXml(src: string): [string, boolean] {
  let out = ''
  for (let i = 0; i < src.length; i++) {
    if ('&' !== src[i]) { out += src[i]; continue }
    const ent = ESC_XML.find((e) => src.startsWith(e[1], i))
    if (undefined === ent) { return ['', false] }
    out += ent[0]
    i += ent[1].length - 1
  }
  return [out, true]
}


function escUri(src: string): string {
  let out = ''
  for (const ch of src) {
    if (1 === ch.length && ESC_URI_UNRESERVED.includes(ch)) {
      out += ch
      continue
    }
    for (const b of utf8Bytes(ch)) {
      out += '%' + b.toString(16).toUpperCase().padStart(2, '0')
    }
  }
  return out
}


function unescUri(src: string): [string, boolean] {
  const bytes: number[] = []
  for (let i = 0; i < src.length; i++) {
    if ('%' !== src[i]) {
      for (const b of utf8Bytes(src[i])) { bytes.push(b) }
      continue
    }
    const hi = hexDigit(src[i + 1] ?? '')
    const lo = hexDigit(src[i + 2] ?? '')
    if (0 > hi || 0 > lo) { return ['', false] }
    bytes.push(hi * 16 + lo)
    i += 2
  }
  return utf8Text(bytes)
}


function escRegex(src: string): string {
  let out = ''
  for (const ch of src) {
    if (ESC_REGEX_PUNCT.includes(ch)) { out += '\\' }
    out += ch
  }
  return out
}


// A metacharacter standing bare is unambiguous and passes; a backslash
// before anything the convention does not escape has no inverse.
function unescRegex(src: string): [string, boolean] {
  let out = ''
  for (let i = 0; i < src.length; i++) {
    if ('\\' !== src[i]) { out += src[i]; continue }
    const n = src[i + 1]
    if (undefined === n || !ESC_REGEX_PUNCT.includes(n)) { return ['', false] }
    out += n
    i++
  }
  return [out, true]
}


function utf8Bytes(ch: string): number[] {
  const c = ch.codePointAt(0) as number
  if (0x80 > c) { return [c] }
  if (0x800 > c) { return [0xC0 | (c >> 6), 0x80 | (c & 0x3F)] }
  if (0x10000 > c) {
    return [0xE0 | (c >> 12), 0x80 | ((c >> 6) & 0x3F), 0x80 | (c & 0x3F)]
  }
  return [
    0xF0 | (c >> 18), 0x80 | ((c >> 12) & 0x3F),
    0x80 | ((c >> 6) & 0x3F), 0x80 | (c & 0x3F),
  ]
}


// The strict decoder: an overlong form, a surrogate, a truncated
// sequence or a value past U+10FFFF has no character to answer with.
function utf8Text(bytes: number[]): [string, boolean] {
  const bad: [string, boolean] = ['', false]
  let out = ''

  for (let i = 0; i < bytes.length; i++) {
    const b = bytes[i]
    if (0x80 > b) { out += String.fromCodePoint(b); continue }

    let need = 0
    let c = 0
    let min = 0
    if (0xC0 === (b & 0xE0)) { need = 1; c = b & 0x1F; min = 0x80 }
    else if (0xE0 === (b & 0xF0)) { need = 2; c = b & 0x0F; min = 0x800 }
    else if (0xF0 === (b & 0xF8)) { need = 3; c = b & 0x07; min = 0x10000 }
    else { return bad }

    if (bytes.length <= i + need) { return bad }
    for (let k = 0; k < need; k++) {
      const cb = bytes[i + 1 + k]
      if (0x80 !== (cb & 0xC0)) { return bad }
      c = (c << 6) | (cb & 0x3F)
    }
    i += need

    if (min > c || 0x10FFFF < c) { return bad }
    if (0xD800 <= c && 0xDFFF >= c) { return bad }
    out += String.fromCodePoint(c)
  }

  return [out, true]
}


// The one entry point each way. An unknown variant never reaches here:
// the call refuses it (`esc_variant`) before either is asked.
function escapeText(src: string, variant: string): string {
  if ('sq' === variant) { return escC(src, "'") }
  if ('sql' === variant) { return escSql(src) }
  if ('shell' === variant) { return escShell(src) }
  if ('xml' === variant) { return escXml(src) }
  if ('uri' === variant) { return escUri(src) }
  if ('regex' === variant) { return escRegex(src) }
  return escC(src, '"')
}


function unescapeText(src: string, variant: string): [string, boolean] {
  if ('sq' === variant) { return unescC(src, "'") }
  if ('sql' === variant) { return unescSql(src) }
  if ('shell' === variant) { return unescShell(src) }
  if ('xml' === variant) { return unescXml(src) }
  if ('uri' === variant) { return unescUri(src) }
  if ('regex' === variant) { return unescRegex(src) }
  return unescC(src, '"')
} /* node:coverage ignore next 8 */


export {
  ESC_VARIANTS,
  isEscVariant,
  escapeText,
  unescapeText,
}

File: go/escape.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

package aontu

import (
	"strings"
	"unicode/utf16"
	"unicode/utf8"
)


// escVariants are the variant names, in the order the reference lists
// them. `none` is not here: it is the absent argument, and the absent
// argument is the C/JSON convention.
var escVariants = []string{"sq", "sql", "shell", "xml", "uri", "regex"}

// escRegexPunct are the regex metacharacters, which are exactly the
// ones the pattern subset admits as escaped-to-mean-themselves
// (reEscapePunct in constraint.go). So esc(s, regex) always answers a
// pattern the subset accepts, which is the point of the variant.
const escRegexPunct = `\.+*?()[]{}|^$/`

var escXml = [][2]string{
	{"&", "&amp;"},
	{"<", "&lt;"},
	{">", "&gt;"},
	{"\"", "&quot;"},
	{"'", "&apos;"},
}

const escUriUnreserved = "ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
	"abcdefghijklmnopqrstuvwxyz0123456789-._~"

const escHexLower = "0123456789abcdef"
const escHexUpper = "0123456789ABCDEF"

func isEscVariant(v string) bool {
	for _, n := range escVariants {
		if n == v {
			return true
		}
	}
	return false
}

// escC is the C escape, in whichever quote the convention uses. Control
// characters below U+0020 with no short form become \uXXXX with
// LOWERCASE hex, as JSON writes them; U+007F is not a JSON escape and
// is left alone.
func escC(src string, quote rune) string {
	var out strings.Builder
	for _, ch := range src {
		switch {
		case '\\' == ch:
			out.WriteString(`\\`)
		case quote == ch:
			out.WriteRune('\\')
			out.WriteRune(quote)
		case '\n' == ch:
			out.WriteString(`\n`)
		case '\r' == ch:
			out.WriteString(`\r`)
		case '\t' == ch:
			out.WriteString(`\t`)
		case '\b' == ch:
			out.WriteString(`\b`)
		case '\f' == ch:
			out.WriteString(`\f`)
		case 0x20 > ch:
			out.WriteString(`\u`)
			for shift := 12; shift >= 0; shift -= 4 {
				out.WriteByte(escHexLower[(ch>>uint(shift))&0xF])
			}
		default:
			out.WriteRune(ch)
		}
	}
	return out.String()
}

// unescC is the inverse. A \uXXXX naming half a surrogate pair is read
// WITH its partner or refused: a lone surrogate is not a character
// either port can carry, and answering one in TypeScript where Go
// cannot is the divergence the whole table exists to avoid.
func unescC(src string, quote rune) (string, bool) {
	var out strings.Builder
	r := []rune(src)

	for i := 0; i < len(r); i++ {
		if '\\' != r[i] {
			out.WriteRune(r[i])
			continue
		}

		i++
		if len(r) <= i {
			return "", false
		}

		switch {
		case '\\' == r[i]:
			out.WriteRune('\\')
			continue
		case quote == r[i]:
			out.WriteRune(quote)
			continue
		case 'n' == r[i]:
			out.WriteRune('\n')
			continue
		case 'r' == r[i]:
			out.WriteRune('\r')
			continue
		case 't' == r[i]:
			out.WriteRune('\t')
			continue
		case 'b' == r[i]:
			out.WriteRune('\b')
			continue
		case 'f' == r[i]:
			out.WriteRune('\f')
			continue
		case 'u' != r[i]:
			return "", false
		}

		hi, ok := hex4(r, i+1)
		if !ok {
			return "", false
		}
		i += 4

		if 0xDC00 <= hi && 0xDFFF >= hi {
			return "", false
		}

		if 0xD800 <= hi && 0xDBFF >= hi {
			// A high surrogate must be followed by its own \u low one.
			if len(r) <= i+2 || '\\' != r[i+1] || 'u' != r[i+2] {
				return "", false
			}
			lo, ok := hex4(r, i+3)
			if !ok || 0xDC00 > lo || 0xDFFF < lo {
				return "", false
			}
			i += 6
			out.WriteRune(utf16.DecodeRune(rune(hi), rune(lo)))
			continue
		}

		out.WriteRune(rune(hi))
	}

	return out.String(), true
}

func hex4(r []rune, at int) (int, bool) {
	if len(r) < at+4 {
		return 0, false
	}
	v := 0
	for i := at; i < at+4; i++ {
		d := hexDigit(r[i])
		if 0 > d {
			return 0, false
		}
		v = v*16 + d
	}
	return v, true
}

func hexDigit(ch rune) int {
	switch {
	case '0' <= ch && '9' >= ch:
		return int(ch - '0')
	case 'a' <= ch && 'f' >= ch:
		return int(ch-'a') + 10
	case 'A' <= ch && 'F' >= ch:
		return int(ch-'A') + 10
	}
	return -1
}

// escSql is standard SQL: a literal's only escape is the doubled quote.
// A LONE quote has no inverse -- it is the escape character, so its
// meaning depends on what was meant to follow it.
func escSql(src string) string {
	return strings.ReplaceAll(src, "'", "''")
}

func unescSql(src string) (string, bool) {
	var out strings.Builder
	r := []rune(src)
	for i := 0; i < len(r); i++ {
		if '\'' != r[i] {
			out.WriteRune(r[i])
			continue
		}
		if len(r) <= i+1 || '\'' != r[i+1] {
			return "", false
		}
		out.WriteRune('\'')
		i++
	}
	return out.String(), true
}

// escShell is POSIX single-quoting: the caller wraps the value in
// single quotes, and a quote inside it closes, escapes and reopens
// them.
func escShell(src string) string {
	return strings.ReplaceAll(src, "'", `'\''`)
}

func unescShell(src string) (string, bool) {
	var out strings.Builder
	r := []rune(src)
	for i := 0; i < len(r); i++ {
		if '\'' != r[i] {
			out.WriteRune(r[i])
			continue
		}
		if len(r) <= i+3 || '\\' != r[i+1] || '\'' != r[i+2] || '\'' != r[i+3] {
			return "", false
		}
		out.WriteRune('\'')
		i += 3
	}
	return out.String(), true
}

func escXmlText(src string) string {
	var out strings.Builder
	for _, ch := range src {
		ent := ""
		for _, e := range escXml {
			if e[0] == string(ch) {
				ent = e[1]
				break
			}
		}
		if "" == ent {
			out.WriteRune(ch)
			continue
		}
		out.WriteString(ent)
	}
	return out.String()
}

// unescXmlText refuses a bare `&` rather than passing it through: it is
// the one character whose meaning is ambiguous here, since it opens an
// entity. Every other character is unambiguous and passes.
func unescXmlText(src string) (string, bool) {
	var out strings.Builder
	for i := 0; i < len(src); {
		if '&' != src[i] {
			out.WriteByte(src[i])
			i++
			continue
		}
		matched := false
		for _, e := range escXml {
			if strings.HasPrefix(src[i:], e[1]) {
				out.WriteString(e[0])
				i += len(e[1])
				matched = true
				break
			}
		}
		if !matched {
			return "", false
		}
	}
	return out.String(), true
}

func escUri(src string) string {
	var out strings.Builder
	for _, b := range []byte(src) {
		if strings.IndexByte(escUriUnreserved, b) >= 0 {
			out.WriteByte(b)
			continue
		}
		out.WriteByte('%')
		out.WriteByte(escHexUpper[b>>4])
		out.WriteByte(escHexUpper[b&0xF])
	}
	return out.String()
}

func unescUri(src string) (string, bool) {
	bytes := make([]byte, 0, len(src))
	for i := 0; i < len(src); i++ {
		if '%' != src[i] {
			bytes = append(bytes, src[i])
			continue
		}
		if len(src) <= i+2 {
			return "", false
		}
		hi := hexDigit(rune(src[i+1]))
		lo := hexDigit(rune(src[i+2]))
		if 0 > hi || 0 > lo {
			return "", false
		}
		bytes = append(bytes, byte(hi*16+lo))
		i += 2
	}
	// The strict decoder: an overlong form, a surrogate, a truncated
	// sequence or a value past U+10FFFF has no character to answer with.
	if !utf8.Valid(bytes) {
		return "", false
	}
	return string(bytes), true
}

func escRegexText(src string) string {
	var out strings.Builder
	for _, ch := range src {
		if strings.ContainsRune(escRegexPunct, ch) {
			out.WriteRune('\\')
		}
		out.WriteRune(ch)
	}
	return out.String()
}

// unescRegexText passes a bare metacharacter, which is unambiguous; a
// backslash before anything the convention does not escape has no
// inverse.
func unescRegexText(src string) (string, bool) {
	var out strings.Builder
	r := []rune(src)
	for i := 0; i < len(r); i++ {
		if '\\' != r[i] {
			out.WriteRune(r[i])
			continue
		}
		if len(r) <= i+1 || !strings.ContainsRune(escRegexPunct, r[i+1]) {
			return "", false
		}
		out.WriteRune(r[i+1])
		i++
	}
	return out.String(), true
}

// escapeText is the one entry point each way. An unknown variant never
// reaches here: the call refuses it (esc_variant) before either is
// asked.
func escapeText(src string, variant string) string {
	switch variant {
	case "sq":
		return escC(src, '\'')
	case "sql":
		return escSql(src)
	case "shell":
		return escShell(src)
	case "xml":
		return escXmlText(src)
	case "uri":
		return escUri(src)
	case "regex":
		return escRegexText(src)
	}
	return escC(src, '"')
}

func unescapeText(src string, variant string) (string, bool) {
	switch variant {
	case "sq":
		return unescC(src, '\'')
	case "sql":
		return unescSql(src)
	case "shell":
		return unescShell(src)
	case "xml":
		return unescXmlText(src)
	case "uri":
		return unescUri(src)
	case "regex":
		return unescRegexText(src)
	}
	return unescC(src, '"')
}
